docs/versioning.md

```diff
@@ -0,0 +1,488 @@
+call.
+
+For example, a standard LIST operation against the keys in a table below would
+return from metadata the list of `[ foo/bar, bar, qux/quz, quz ]` using the
+separator represented by `.` as the extra delimiter.
+
+| key |
+|-----|
+| foo/bar |
+| foo/bar.v2 |
+| foo/bar.v1 |
+| bar |
+| qux/quz |
+| qux/quz.v2 |
+| qux/quz.v1 |
+| quz |
+| quz.v2 |
+| quz.v1 |
+
+## Implementation of Bucket Versioning in API
+
+### Object Metadata Versioning Attributes
+
+To access all the information needed to properly handle all cases that may exist
+in versioned operations, the API stores certain versioning-related information
+in the metadata attributes of each version's object metadata.
+
+These are the versioning-related metadata properties:
+
+- `isNull`: whether the version being stored is a null version.
+- `nullVersionId`: the version ID of the latest null version.
+- `isDeleteMarker`: whether the version being stored is a delete marker.
+- `versionId`: the version ID of the version being stored.
+
+Null versions and delete markers are described in further detail in the
+following sections.
+
+### Creation of New Versions
+
+When versioning is enabled in a bucket, APIs which normally result in the
+creation of objects, such as Put Object, Complete Multipart Upload and Copy
+Object, will generate new versions of objects.
+
+S3 creates a new version and updates the master version using the
+`versioning: true` option in PUT calls to the metadata engine. As an example,
+when two consecutive Put Object requests are sent to the S3 Server for a
+versioning-enabled bucket with the same key names, there are two corresponding
+metadata PUT calls with the `versioning` option set to true.
+
+The PUT calls to metadata and resulting keys are shown below:
+
+(1) PUT foo (first put), versioning: `true`
+
+| key | value |
+|-----|:-----:|
+| foo |   A   |
+| foo.v1 |   A   |
+
+(2) PUT foo (second put), versioning: `true`
+
+| key | value |
+|-----|:-----:|
+| foo |   B   |
+| foo.v2 |   B   |
+| foo.v1 |   A   |
+
+#### Null Version Management
+
+In a bucket without versioning, or when versioning is suspended, putting an
+object with the same name twice should result in the previous object being
+overwritten. This is managed with null versions.
+
+Null versions are generated when an object is put before versioning is enabled
+or when versioning is suspended, and are distinguished from normal versions in
+that they are overwritten by new null versions. Therefore, only one null version
+should exist at any given time, and it can be identified in S3 requests
+and responses with the version id "null".
+
+With respect to metadata, since the null version is overwritten by subsequent
+null versions, the null version is initially stored in the master key alone,
+as opposed to being stored in the master key and a new version, like how normal
+versions are stored. The metadata engine does not differentiate between null and
```

Comment:
I'd get rid of "like how normal versions are stored."  Seem unnecessary and a bit confusing.